File: 2020/question_2.py
# Score: 11/15

class Cell:
    def __init__(self, r, c): # Default constructor
        self.row = r # Variables are public by default
        self.col = c

checked = [[False for i in range(1000)] for j in range(1000)] # 2D array with values as False
values = [[] for i in range(1000000)] # 2D array with all of the cells

# Possible Function
def possible(prev_cell):
    queue = [] # Initialize a queue
    queue.append(prev_cell) # Adds previous element to the queue

    while len(queue) > 0:
        temp_cell = queue[0] # Gets value of first element in the queue
        queue.pop(0) # Removes first element in queue at specified location
        jump_possibilities = values[temp_cell.row * temp_cell.col] # Array of all cells that can be jumped to

        for cell in jump_possibilities:
            if cell.row == 1 and cell.col == 1: # If in top left corner, then exit is possible
                return "yes" # END FUNCTION
            elif not(checked[cell.row][cell.col]): # If we have not searched the cell before
                checked[cell.row][cell.col] = True
                queue.append(cell) # Add cell to queue

    return "no" # If no escape is found

# Main function
rows = int(input())
cols = int(input())

prev_cell = Cell(0, 0)

for i in range(1, rows + 1):
    numbers = list(map(int, input().split())) # Makes an array of integer inputs
    for j in range(1, cols + 1):
        temp = Cell(i, j)
        values[numbers[j - 1]].append(temp)

        if i == rows and j == cols:
            prev_cell = temp

print(possible(prev_cell)) # answer

File: 2020/question_2.cpp
/* Score: 13/15 */

#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

class Cell { //cell class will represent all of the present cells
public:
    int row;
    int col;

    Cell() {
        row = 0;
        col = 0;
    }

    Cell(int r, int c) {
        row = r;
        col = c;
    }
};

//1000 was chosen because it is the maximum values for M and N (range is 1 to 1000 inclusive)
bool checked[1000][1000]; //keep track if cells have been checked or now
vector<vector<Cell>> vals(1000000); //values of the cells can range from 1 to 1000000. If a cell has a specific value (say 10), it will be stored in the child vector at the tenth element of the parent vector

string possible(Cell prev_cell) {
    queue<Cell> cells;
    cells.push(prev_cell); //adds prev_cell to the end of the queue

    while(cells.size() > 0) {
        Cell tempCell = cells.front();
        cells.pop();
        vector<Cell> jumpPossibilities = vals[tempCell.row * tempCell.col]; //gets vector of all cells that have the correct value

        for(Cell jump: jumpPossibilities) {
            if(jump.row == 1 && jump.col == 1) { //if cell is at the top-left corner, then we have escaped :))
                return "yes"; //returns "yes" and EXITS FUNCTION
            } else if(!checked[jump.row][jump.col]) { //if the cell has not been searched yet... 
                checked[jump.row][jump.col] = true; 
                cells.push(jump); //... we add it to the back of the queue and eventually, as the while loop loops, it will use this cell as a reference to look for paths to escape
            }
        }
    }
    
    return "no"; //if the size of the cells queue becomes zero, it means no escape was possible
}

int main() {

    //gets number of rows and columns
    int rows = 0, columns = 0;
    cin >> rows >> columns;

    Cell prev_cell;

    int input = 0;
    //coordinates start at 1, so the counter starts at 1 as well
    for(int i = 1; i <= rows; i++) {
        for(int j = 1; j <= columns; j++) {
            cin >> input; //value of the cell
            Cell temp(i, j); //creates a cell object
            vals[input].push_back(temp); //adds cell

            if(i == rows && j == columns) {
                prev_cell = temp; //sets the starting cell at bottom right corner
            }
        }
    }

    cout << possible(prev_cell); //calls function to see if there is a possibility for escape
    return 0;
}